File: src/main/java/com/kdopamine/app/domain/member/entity/Member.java
package com.kdopamine.app.domain.member.entity;

import com.kdopamine.app.global.entity.BaseEntity;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;

@Entity
@Table(name = "members")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder(access = AccessLevel.PRIVATE)
@Getter
public class Member extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 100, unique = true)
    private String email;

    @Column(nullable = false, length = 100)
    private String nickname;

    @Column(nullable = false, length = 10)
    private String provider;   // GOOGLE, APPLE

    @Column(nullable = false)
    private String providerId; // 소셜 서비스에서 제공하는 고유 식별값 (PK 역할)

    @Enumerated(EnumType.STRING)
    private Role role; // USER, ADMIN

    @Column(nullable = false)
    private Integer level = 1;

    @Column(nullable = false)
    private Integer currentXp = 0;

    @Column(nullable = false)
    private Integer streakDays = 0;

    @Column(nullable = false)
    private Integer totalAttendanceDays = 0;

    private LocalDate lastAttendanceDate;

    // 소셜 로그인 신규 회원 생성용
    public static Member createSocialMember(String email, String nickname, String provider, String providerId) {
        return Member.builder()
                .email(email)
                .nickname(nickname)
                .provider(provider)
                .providerId(providerId)
                .role(Role.USER)
                .build();
    }

    public void gainXp(int amount) {
        this.currentXp += amount;

        while (this.currentXp >= getRequiredXpForNextLevel()) {
            this.currentXp -= getRequiredXpForNextLevel();
            this.level++;
        }
    }

    public int getRequiredXpForNextLevel() {
        return (int) (100 * Math.pow(1.2, this.level - 1));
    }
}

File: k_lingo_front/lib/models/member.dart
// 사용자 역할
enum Role { USER, ADMIN }

// 소셜 로그인 제공자
enum Provider { GOOGLE, APPLE }

class Member {
  final int? id;
  final String email;
  final String nickname;
  final String provider;   // GOOGLE, APPLE
  final String providerId; // 소셜 서비스에서 제공하는 고유 식별값
  final Role role;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Member({
    this.id,
    required this.email,
    required this.nickname,
    required this.provider,
    required this.providerId,
    this.role = Role.USER,
    this.createdAt,
    this.updatedAt,
  });

  factory Member.fromJson(Map<String, dynamic> json) {
    return Member(
      id: json['id'],
      email: json['email'] ?? '',
      nickname: json['nickname'] ?? '',
      provider: json['provider'] ?? '',
      providerId: json['providerId'] ?? '',
      role: _parseRole(json['role']),
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : null,
      updatedAt: json['updatedAt'] != null 
          ? DateTime.parse(json['updatedAt']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'email': email,
      'nickname': nickname,
      'provider': provider,
      'providerId': providerId,
      'role': role.name,
    };
  }

  static Role _parseRole(String? role) {
    switch (role) {
      case 'ADMIN':
        return Role.ADMIN;
      default:
        return Role.USER;
    }
  }

  // 편의 메서드
  bool get isAdmin => role == Role.ADMIN;
  bool get isGoogleUser => provider == 'GOOGLE';
  bool get isAppleUser => provider == 'APPLE';
}
